src/compiler/utilities.ts

```diff
@@ -8253,4 +8255,60 @@ namespace ts {
         // '/// <reference no-default-lib="true"/>' directive.
         return options.skipLibCheck && sourceFile.isDeclarationFile || options.skipDefaultLibCheck && sourceFile.hasNoDefaultLib;
     }
+
+    // TODO: remove once tslint allows tsconfig.json to include lib "esnext.bigint"
+    /* @internal */
+    declare const BigInt: ((value: number | string) => number) | undefined;
+
+    // Converts a bigint literal, e.g. `0x1234n`,
+    // to its decimal representation, e.g. `4660`.
+    export function parsePseudoBigInt(stringValue: string): string {
+        // Use native BigInt if available
+        if (typeof BigInt !== "undefined") {
+            return "" + BigInt(stringValue.slice(0, -1)); // omit trailing "n"
+        }
+
+        let log2Base: number;
+        if (stringValue.startsWith("0b")) log2Base = 1;
+        else if (stringValue.startsWith("0o")) log2Base = 3;

```

Comment:
These base indicator comparisons need to be case-insensitive. `0O` and `0X` are afaik all still allowed on bigints just like normal ints. personally, I think slicing the first two characters and switching on them to determine the base is a bit nicer looking, but that's a matter of preference.

Reply:
`0b`, `0o`, and `0x` are not parsed directly from the code; the scanner always lower-cases the base prefix. See https://github.com/Microsoft/TypeScript/pull/25886/files/2edab35e278984ae7f30ff28a2698a64a7e390d5#diff-a15280067dcb3b0929808762f8352e86R1572. I'll add tests for it though.

Reply:
Right, but there's nothing enforcing that invariant in the scanner and it's reasonable to expect that someone calls this with the raw text of a literal at some point.